### DISCUSSION TITLE
QGIS cannot locate procedure entry point after activation of plugin

### DISCUSSION THREAD

**Discussion by EddyGC61** [coolzhao/Geo-SAM#61] [Q&A]
Hi, I attempted to install Release 1.3.1 RC3 in accordance with the instructions. It all seemed to go very well until I checked on the plugin to activate it in QGis.

I received the error (see screen grab) it could not locate a procedure entry point and the following error description.

I am not a coder, but an end user of QGIS. Could I have missed something in the installation? Any help would be appreciated.

![GeoSAMnoentrypoint](https://github.com/user-attachments/assets/20e03132-1bde-437e-b849-7b61fdb86a3d)

**Couldn't load plugin 'Geo-SAM'** 

**rasterio.errors.CRSError: The EPSG code is unknown. PROJ: proj_create_from_database: C:\OSGeo4W\share\proj\proj.db contains DATABASE.LAYOUT.VERSION.MINOR = 3 whereas a number >= 4 is expected. It comes from another PROJ installation.** 
Traceback (most recent call last):
  File "rasterio\\crs.pyx", line 592, in rasterio.crs.CRS.from_epsg
  File "rasterio\\_err.pyx", line 289, in rasterio._err.exc_wrap_int
rasterio._err.CPLE_AppDefinedError: PROJ: proj_create_from_database: C:\OSGeo4W\share\proj\proj.db contains DATABASE.LAYOUT.VERSION.MINOR = 3 whereas a number >= 4 is expected. It comes from another PROJ installation.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\OSGeo4W/apps/qgis-ltr/./python\qgis\utils.py", line 401, in loadPlugin
    __import__(packageName)
  File "C:\OSGeo4W/apps/qgis-ltr/./python\qgis\utils.py", line 892, in _import
    mod = _builtin_import(name, globals, locals, fromlist, level)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users/biost/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\Geo-SAM\__init__.py", line 7, in 
    from .geo_sam_tool import Geo_SAM
  File "C:\OSGeo4W/apps/qgis-ltr/./python\qgis\utils.py", line 892, in _import
    mod = _builtin_import(name, globals, locals, fromlist, level)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users/biost/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\Geo-SAM\geo_sam_tool.py", line 10, in 
    from .tools.widgetTool import Selector, EncoderCopilot
  File "C:\OSGeo4W/apps/qgis-ltr/./python\qgis\utils.py", line 892, in _import
    mod = _builtin_import(name, globals, locals, fromlist, level)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users/biost/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\Geo-SAM\tools\widgetTool.py", line 21, in 
    from torchgeo.datasets import BoundingBox
  File "C:\OSGeo4W/apps/qgis-ltr/./python\qgis\utils.py", line 892, in _import
    mod = _builtin_import(name, globals, locals, fromlist, level)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\OSGeo4W\apps\Python312\Lib\site-packages\torchgeo\datasets\__init__.py", line 6, in 
    from .advance import ADVANCE
  File "C:\OSGeo4W/apps/qgis-ltr/./python\qgis\utils.py", line 892, in _import
    mod = _builtin_import(name, globals, locals, fromlist, level)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\OSGeo4W\apps\Python312\Lib\site-packages\torchgeo\datasets\advance.py", line 19, in 
    from .geo import NonGeoDataset
  File "C:\OSGeo4W/apps/qgis-ltr/./python\qgis\utils.py", line 892, in _import
    mod = _builtin_import(name, globals, locals, fromlist, level)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\OSGeo4W\apps\Python312\Lib\site-packages\torchgeo\datasets\geo.py", line 47, in 
    class GeoDataset(Dataset[dict[str, Any]], abc.ABC):
  File "C:\OSGeo4W\apps\Python312\Lib\site-packages\torchgeo\datasets\geo.py", line 90, in GeoDataset
    _crs = CRS.from_epsg(4326)
           ^^^^^^^^^^^^^^^^^^^
  File "rasterio\\crs.pyx", line 596, in rasterio.crs.CRS.from_epsg
rasterio.errors.CRSError: The EPSG code is unknown. PROJ: proj_create_from_database: C:\OSGeo4W\share\proj\proj.db contains DATABASE.LAYOUT.VERSION.MINOR = 3 whereas a number >= 4 is expected. It comes from another PROJ installation.

Python version: 3.12.6 (main, Sep  7 2024, 18:59:42) [MSC v.1938 64 bit (AMD64)] 
QGIS version: 3.34.11-Prizren Prizren, 2904bcec 

Python Path:
C:/OSGeo4W/apps/qgis-ltr/./python
C:/Users/biost/AppData/Roaming/QGIS/QGIS3\profiles\default/python
C:/Users/biost/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins
C:/OSGeo4W/apps/qgis-ltr/./python/plugins
C:\OSGeo4W\apps\grass\grass84\etc\python
C:\WINDOWS\system32
C:\OSGeo4W\bin\python312.zip
C:\OSGeo4W\apps\Python312\DLLs
C:\OSGeo4W\apps\Python312\Lib
C:\OSGeo4W\bin
C:\OSGeo4W\apps\Python312
C:\OSGeo4W\apps\Python312\Lib\site-packages
C:\OSGeo4W\apps\Python312\Lib\site-packages\win32
C:\OSGeo4W\apps\Python312\Lib\site-packages\win32\lib
C:\OSGeo4W\apps\Python312\Lib\site-packages\Pythonwin
C:/Users/biost/AppData/Roaming/QGIS/QGIS3\profiles\default/python

**Comment by Fanchengyan** [comment]
Hi @EddyGC61 

This is the same issue as https://github.com/coolzhao/Geo-SAM/issues/57#issuecomment-2550531518. You can try applying the solution suggested there to see if it resolves the problem.

**Reply by EddyGC61** [reply]
Hi @Fanchengyan 

_> Hi @EddyGC61
> 
> This is the same issue as [#57 (comment)](https://github.com/coolzhao/Geo-SAM/issues/57#issuecomment-2550531518). You can try applying the solution suggested there to see if it resolves the problem._

Thank you for that assist. I did try as you suggested, but with the same error resulting. 

What is the process of manually updating the proj.db? I viewed the file and it is not text format. Do I overwrite with a file from elsewhere, and where would I obtain that file?

Again, thank you for your assistance.

**Comment by Fanchengyan** [comment]
Hi @EddyGC61 

The issue stems from having multiple GIS software installations, which can result in multiple versions of PROJ being installed and lead to inconsistencies in the `proj.db` files. I haven’t encountered this issue myself, so I don’t have direct experience solving it. Below are a few potential fixes you can try:

1. **Search for Multiple `proj.db` Files**  
   Use the [Everything](https://www.voidtools.com/) search tool to locate all `proj.db` files on your system. If you find multiple instances, try setting the `PROJ_DATA` environment variable to point to a another `proj.db` file and see if that resolves the issue.

2. **Reinstall `pyproj` and `rasterio`**  
   Delete the existing `proj.db` file, then reinstall `pyproj` and `rasterio` with the following command:

   ```bash
   pip install pyproj rasterio --force-reinstall
   ```

3. **Uninstall Other GIS Software**  
   If you have multiple GIS applications installed, try uninstalling all of them, then reinstall `pyproj` and `rasterio` using the command above.

4. **Install QGIS via Conda**  
   If the issue persists, try installing QGIS via Conda and check whether the Conda version of QGIS encounters the same problem. For installation and usage instructions, refer to this guide: [QGIS Installation with Conda](https://geo-sam.readthedocs.io/en/latest/installation.html#install-qgis).

If you manage to resolve the issue, please share your solution with the community to help others facing similar problems. Good luck!

**Reply by EddyGC61** [reply]
Thank you for the extra information. I attempted all your suggestions without luck. I have fired up an older PC to see if I am able to get it to work on that. Thank you again.

**Reply by coolzhao** [reply]
Hi @EddyGC61, this seems to be an installation issue of QGIS. You can try to clean reinstall following a complete uninstallation. 
You can refer to the following link to uninstall. https://courses.spatialthoughts.com/uninstall-qgis-ltr.html
Good luck.

**Reply by EddyGC61** [reply]
I've already tried a clean install (uninstalled using OSGeo4W and then removed everything I could find relating to QGIS and OSGeo4W including references within the registry) and still got the same error message. 

It will be something simple and easily overlooked. 

PS. I couldn't get it to work on another PC (also after a clean install). The two PCs have similar software loaded, and I did notice a proj.db in the Jamovi directory on the second PC which I removed with every other copy I could find after removing QGIS from the system, installed QGIS from OSGeo4W, reinstalled pyproj and rasterio, installed Geo-SAM and got the same error message.

**Reply by coolzhao** [reply]
Got it. We will try to figure out what the problem is here. Thanks for your feedback and detailed information.

**Reply by EddyGC61** [reply]
Thanks for your help. I hope you are able to resolve it soon.
